Tasks/Gradle/sonarqube-common/sonarqube-rest.ts

```diff
@@ -0,0 +1,193 @@
+            })
+    }, delay * 1000, taskReport);
+
+    // After [timeout] seconds, delete the repeating call and return false.
+    setTimeout((intervalObject) => {
+        clearInterval(intervalObject);
+        defer.resolve(false);
+    }, timeout * 1000, intervalObject);
+
+    return defer.promise;
+}
+
+// Queries the server to determine if the task has finished, i.e. if the quality gate has been evaluated
+export function isTaskComplete(taskReport:TaskReport):Q.Promise<boolean> {
+    return getTaskDetails(taskReport)
+        .then((responseJson:any) => {
+            var taskStatus:string = responseJson.task.status;
+            return (taskStatus.toUpperCase() == 'SUCCESS');
+        });
+}
+
+// Query the server to determine the analysis id associated with the current task
+export function getAnalysisId(taskReport:TaskReport):Q.Promise<string> {
+    return getTaskDetails(taskReport)
+        .then((analysisDetails:any) => {
+            return getTaskAnalysisId(analysisDetails);
+        });
+}
+
+function getTaskAnalysisId(analysisDetails:any):string {
+    return analysisDetails.task.analysisId;
+}
+
+// Returns the status of the analysis identified by the argument.
+export function getAnalysisStatus(analysisId:string):Q.Promise<string> {
+    return getAnalysisDetails(analysisId)
+        .then((analysisDetails:any) => {
+            return getProjectStatus(analysisDetails);
+        });
+}
+
+function getProjectStatus(qualityGateDetails:any):string {
+    return qualityGateDetails.projectStatus.status;
+}
+
+export function getAnalysisDetails(analysisId:string):Q.Promise<Object> {
+    try {
+        return callSonarQubeRestEndpoint('/api/qualitygates/project_status?analysisId=' + analysisId)
+            .then((response:RestResponse) => {
+                var responseJson:any = response.payloadToJson();
+                if (!responseJson.projectStatus) {
+                    throw responseJson;
+                }
+
+                return responseJson;
+            });
+    } catch (error) {
+        tl.debug('Could not fetch quality gate details on analysis ID' + analysisId);
+        throw error;
+    }
+}
+
+function getTaskDetails(taskReport:TaskReport):Q.Promise<Object> {
+    if (!taskReport) {
+        throw new Error(tl.loc('sqAnalysis_TaskReportInvalid'));
+    }
+
+    try {
+        return callSonarQubeRestEndpoint('/api/ce/task?id=' + taskReport.ceTaskId)
+            .then((response:RestResponse) => {
+                var responseJson:any = response.payloadToJson();
+                if (!responseJson.task) {
+                    throw responseJson;
+                }
+
+                return responseJson;
+            });
+    } catch (error) {
+        tl.debug('Could not fetch task details on ID' + taskReport.ceTaskId);
+        throw error;
+    }
+}
+
+// Invokes a REST endpoint at the SonarQube server.
+function callSonarQubeRestEndpoint(path:string):Q.Promise<RestResponse> {
```

Comment:
I don't think there's any point in using this RestResponse object. If the response is not successful, the promise should be rejected. If the response is successful, deserialize it. If the deserialization fails, you also need to reject it.